### PR TITLE
Basic Alonzo node

### PR DESCRIPTION
**Pull request by Jimbo4350**
In order to have QA begin testing on an Alonzo capable node we have minimally integrated the Alonzo era such that a node can hardfork to the Alonzo era. No other functionality has been implemented i.e nothing cli related and only enough in the api to allow a hardfork.

NB: #2400 needs to be merged to enable hardforking to Alonzo

### CODE REVIEW

**Comment by Jimbo4350** [comment]
After 2932 in ouroboros-network is merged, I can update the node deps in this PR and finalize this PR.

File: cardano-api/src/Cardano/Api/Fees.hs [hunk 1]
```diff
@@ -42,17 +42,13 @@ transactionFee :: forall era.
...
--- and remove the IsShelleyBasedEra constraint.
-transactionFee _ _ (ByronTx _) =
-    case shelleyBasedEra :: ShelleyBasedEra era of {}
+transactionFee txFeeFixed txFeePerByte tx =
+  let a = toInteger txFeePerByte
+      b = toInteger txFeeFixed
+  in case tx of
+       ShelleyTx _ tx' -> let x = getField @"txsize" tx'
+                          in Lovelace (a * x + b)
+       ByronTx _ -> case shelleyBasedEra :: ShelleyBasedEra ByronEra of {}
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Fees.hs:51]: This was just a refactor

File: cardano-api/src/Cardano/Api/Tx.hs [hunk 2]
```diff
@@ -134,6 +134,8 @@ instance Eq (Tx era) where
         ShelleyBasedEraShelley -> txA == txB
         ShelleyBasedEraAllegra -> txA == txB
         ShelleyBasedEraMary    -> txA == txB
+        ShelleyBasedEraAlonzo  ->
+          error "Eq (Tx era): Alonzo era not implemented yet"
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Tx.hs:138]: They do give us the `Eq` instance, this was me just mechanically filling in Alonzo things.

File: cardano-api/src/Cardano/Api/Tx.hs [hunk 3]
```diff
@@ -159,6 +161,10 @@ instance Show (Tx era) where
         showString "ShelleyTx ShelleyBasedEraMary "
       . showsPrec 11 tx
 
+    showsPrec _ (ShelleyTx ShelleyBasedEraAlonzo _) =
+      error "Show (Tx era): Alonzo era not implemented yet"
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Tx.hs:165]: Same as above. I was mechanically filling things in.

File: cardano-node/cardano-node.cabal [hunk 4]
```diff
@@ -139,6 +140,7 @@ library
                       , safe-exceptions
                       , scientific
                       , shelley-spec-ledger
+                      , small-steps
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/cardano-node.cabal:146]: This is fore `Data.MemoBytes` for a couple of the orphan json instances e.g: `deriving instance ToJSON (MemoBytes.MemoBytes (Map ByteString Integer))`

File: cardano-node/src/Cardano/Node/Orphans.hs [hunk 5]
```diff
@@ -26,6 +45,34 @@ instance FromJSON TracingVerbosity where
...
+--TODO: I assume we are not interested in rendering
+--plutus scripts as JSON.
+instance ToJSON SBS.ShortByteString where
+  toJSON _plutusScriptBytes = Aeson.String "Plutus script placeholder"
+
+-- Obviously incorrect. Need to fix.
+deriving instance ToJSONKey SBS.ShortByteString
+-- Obviously incorrect. Need to fix.
+deriving instance ToJSONKey ByteString
+deriving instance ToJSONKey Alonzo.Language
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Orphans.hs:64]: If you are talking about `deriving instance ToJSONKey Alonzo.Language` then yes I believe so. `Language` is a simple ADT: `data Language = PlutusV1 `

File: cardano-node/src/Cardano/Node/Orphans.hs [hunk 6]
```diff
@@ -26,6 +45,34 @@ instance FromJSON TracingVerbosity where
...
+deriving instance ToJSON Alonzo.CostModel
+
+
+deriving instance ToJSON (Alonzo.TxOut (Alonzo.AlonzoEra StandardCrypto))
+
+deriving instance ToJSON (Shelley.CompactAddr StandardCrypto)
+deriving instance Generic (Shelley.CompactAddr StandardCrypto)
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Orphans.hs:71]: :+1:  cc: @nc6

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 7]
```diff
@@ -164,6 +168,15 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
...
             ProtVer 4 0
         }
+        Consensus.ProtocolParamsAlonzo {
+          -- This is /not/ the Mary protocol version. It is the protocol
+          -- version that this node will declare that it understands, when it
+          -- is in the Mary era. Since Mary is currently the last known
+          -- protocol version then this is also the Mary protocol version.
+          alonzoProtVer =
+            ProtVer 4 0,
+          alonzoTranslationContext = 42
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Protocol/Cardano.hs]: This is an outdated commit based on an earlier commit of `ouroboros-network`

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 8]
```diff
@@ -164,6 +168,15 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
           maryProtVer =
             ProtVer 4 0
         }
+        Consensus.ProtocolParamsAlonzo {
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Protocol/Cardano.hs:182]: :+1:

File: cardano-node/src/Cardano/Node/Protocol/Alonzo.hs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+readAlonzoGenesis
+  :: FilePath
+  -> ExceptT AlonzoProtocolInstantiationError IO Alonzo.AlonzoGenesis
+readAlonzoGenesis fpath = do
+  alonzoGenWrapper <- readAndDecode
+                       `catchError` \err ->
+                         case err of
+                           AlonzoGenesisFileError (FileIOError _ ioe)
+                             | isDoesNotExistError ioe -> panic "Shelley genesis file not found."
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Protocol/Alonzo.hs:54]: Yep the `panic` was a placeholder. I'll propagate and render the error correctly.

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 10]
```diff
@@ -171,12 +180,12 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
         Consensus.ProtocolParamsAlonzo {
           -- This is /not/ the Mary protocol version. It is the protocol
           -- version that this node will declare that it understands, when it
-          -- is in the Mary era. Since Mary is currently the last known
-          -- protocol version then this is also the Mary protocol version.
-          alonzoProtVer =
-            ProtVer 4 0,
-          alonzoTranslationContext = 42
+          -- is in the Alonzo era. Since Alonzo is currently the last known
+          -- protocol version then this is a        Consensus.ProtocolParamsTransition {
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Protocol/Cardano.hs:186]: Yep copy and paste error

File: cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs [hunk 11]
```diff
@@ -125,3 +134,22 @@ instance ToObject ResourceStats where
...
+    => ToJSON (TxOut era) where
+  toJSON (Alonzo.TxOut addr v dataHash) =
+    object [ "address" .= toJSON addr
+           , "value" .= toJSON v
+           , "datahash" .= case strictMaybeToMaybe dataHash of
+                             Nothing -> Aeson.Null
+                             Just dHash ->
+                               Aeson.String . Crypto.hashToTextAsHex
+                                 $ SafeHash.extractHash dHash
+           ]
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Tracing/OrphanInstances/Common.hs:148]: Yep this should

File: cardano-cli/src/Cardano/CLI/Shelley/Orphans.hs [hunk 12]
```diff
@@ -95,3 +99,27 @@ deriving newtype  instance ToJSON    (Ledger.Mary.PolicyID StandardCrypto)
...
+
+
+
+deriving instance FromJSON Alonzo.ExUnits
+deriving instance FromJSON Alonzo.Prices
+
+-- We defer parsing of the cost model so that we can
+-- read it as a filepath. This is to reduce further pollution
+-- of the genesis file.
+instance FromJSON Alonzo.AlonzoGenesis where
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Orphans.hs:111]: Yep it is

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs [hunk 13]
```diff
@@ -919,3 +930,67 @@ runGenesisHashFile (GenesisFile fpath) = do
    let gh :: Crypto.Hash Crypto.Blake2b_256 ByteString
        gh = Crypto.hashWith id content
    liftIO $ Text.putStrLn (Crypto.hashToTextAsHex gh)
+
+--
+-- Alonzo genesis
+--
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs:936]: Yep that's correct

**Review by dcoutts — CHANGES_REQUESTED** [review]
Nearly there.

File: cardano-node/src/Cardano/Node/Orphans.hs [hunk 14]
```diff
@@ -55,11 +57,14 @@ deriving instance ToJSON (MemoBytes.MemoBytes (Map ByteString Integer))
 --TODO: I assume we are not interested in rendering
 --plutus scripts as JSON.
 instance ToJSON SBS.ShortByteString where
-  toJSON _plutusScriptBytes = Aeson.String "Plutus script placeholder"
+  toJSON plutusScriptBytes =
+    case Base16.decode $ SBS.fromShort plutusScriptBytes of
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Node/Orphans.hs:61]: Silly slip up on my part

**Review by dcoutts — APPROVED** [review]

File: cardano-node/src/Cardano/Node/Orphans.hs (hunk 14 above)
[the remark this answers is not in the document]
Comment (dcoutts) [reply cardano-node/src/Cardano/Node/Orphans.hs]: I've simplified these ToJSON instances now so this instance is not needed at all.

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 15]
```diff
@@ -171,12 +172,17 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
           -- version that this node will declare that it understands, when it
           -- is in the Mary era. Since Mary is currently the last known
           -- protocol version then this is also the Mary protocol version.
+          --
+          -- During testing of the Alonzo era, we conditionally declare that we
+          -- know about the Alonzo era. We do so only when a config option for
+          -- testing development/unstable eras is used. This lets us include
+          -- not-yet-ready eras in released node versions without mainnet nodes
```
Comment (cblp) [inline cardano-node/src/Cardano/Node/Protocol/Cardano.hs:179]: Can a specially forged packet break a node?

File: scripts/byron-to-alonzo/mkfiles.sh [hunk 16]
```diff
@@ -1,7 +1,10 @@
 #!/usr/bin/env bash
 
-set -e
```
Comment (cblp) [inline scripts/byron-to-alonzo/mkfiles.sh:3]: `-e` is still useful

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 17]
```diff
@@ -974,6 +1009,7 @@ instance IsCardanoEra era => SerialiseAsCBOR (TxBody era) where
                         (ShelleyTxBody ShelleyBasedEraAllegra) bs
         MaryEra    -> deserialiseShelleyBasedTxBody
                         (ShelleyTxBody ShelleyBasedEraMary) bs
+        AlonzoEra    -> error "deserialiseFromCBOR: Alonzo era not implementd yet"
```
Comment (cblp) [inline cardano-api/src/Cardano/Api/TxBody.hs:1012]: "implemented"

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs (hunk 15 above)
Comment (dcoutts) [reply cardano-node/src/Cardano/Node/Protocol/Cardano.hs:179]: No. It's just about not prematurely advertising support for a new era.

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 18]
```diff
@@ -229,3 +262,6 @@ renderCardanoProtocolInstantiationError
 renderCardanoProtocolInstantiationError
   (CardanoProtocolInstantiationErrorShelley err) =
     Shelley.renderShelleyProtocolInstantiationError err
+
+renderCardanoProtocolInstantiationError
+  (CardanoProtocolInstantiationErrorAlonzo _) = error "FIX ME"
```
Comment (cblp) [inline cardano-node/src/Cardano/Node/Protocol/Cardano.hs:273]: why not "not implemented yet"?

**Comment by Jimbo4350** [comment]
bors merge

**Pull request merged**